README.md
# Challenge_TextMine_2023 :  "Reconnaissance d’entités d’intérêts dans les signatures d’e-mails"

# Key informations:

1. Challenge details: https://github.com/Emvista/Challenge_TextMine_2023
2. Result deadline: **9/01/2023** (Test Data : **8/01/2023**)   
3. Paper deadline (2 pages): **13/01/2023**
4. Result submission: [textmine@emvista.com](mailto:textmine@emvista.com). **3 submissions max**
5. Paper submission: [https://easychair.org/conferences/?conf=textmine23](https://easychair.org/conferences/?conf=textmine23)

# Dataset: 

- **Authentique** (JDA) : signatures authentiques pseudonymisées ; non disponible pour l'instant (jeu de test sur lequel seront évalués les participants au défi)
- **Réaliste** (JDR) : signatures construites manuellement par la société Isahit, plateforme de labellisation éthique des données pour l'IA ; ce jeu de données contient des signatures réalistes, c'est-à-dire proches des signatures authentiques observées, mais non authentiques : https://github.com/Emvista/Challenge_TextMine_2023/blob/main/JDR.json
- **Factice** (JDF) : signatures créées automatiquement à partir d'une API de génération de fausses identités ; disponible ici : https://github.com/Emvista/Challenge_TextMine_2023/blob/main/JDF.json

# Participants:

- Nicolas Fouqué
- Hugo Thomas
- Cyrielle Mallart
- Duc Hau Nguyen
- Victor Klötzer

# TODO:

- [x] Responsable de communication: **Nico**
- [x] Etat de l'art: **Hugo**
- [ ] GCN: **Hugo**
- [ ] CRF: **Hugo**
- [x] Bonsai Boost: **Nico**
- [ ] BERT-NERT: **Duc**
  - [x] Camembert-BERT vanilla test
  - [ ] Variation de tokenization
  - [ ] Retrain on a limited number of classes
    - [ ] Function, Organization
    - [ ] Function, Organization, Project
    - [ ] Function, Organization, Project, Location
- [x] RegExp: **Victor**
- [ ] QA : **Cyrielle**
- [ ] Luke: **Cyrielle**
- [ ] Canal de communication: Slack

# Result




dataset/CHANGELOG.md
# CHANGELOG

## 30 Décembre 2022

* Correction de formes incorrectes (qui terminent par \n) dans le JDF

## 27 Décembre 2022

* Homogénéisation des formats : les sauts de lignes du JDR sont désormais encodées "\n" plutôt que "\\r\\n"
* Correction d'une annotation fausse dans le JDR

## 16 Décembre 2022

* Correction d'une erreur d'offsets dans le JDR
* Correction d'une annotation incomplète dans le JDR

## 20 Octobre 2022

* Version initiale


merge_results.py
import json

CLASSES_REGEX = ['Email', 'Phone_Number', 'Reference_CEDEX', 'Reference_CS', 'Reference_Code_Postal', 'Social_Network',
                 'Url', 'Reference_User']

REGEX_DATA = 'annotations/preds_regex.json'
CRF_DATA = 'annotations/preds_crf.json'
BONZAI_DATA = 'annotations/preds_bonzai.json'
BERT_DATA = 'annotations/preds_bert.json'

if __name__ == '__main__':
    # print("Loading Predictions")
    with open(REGEX_DATA) as in_f:
        regex_data = json.load(in_f)
    with open(CRF_DATA) as in_f:
        crf_data = json.load(in_f)
    with open(BONZAI_DATA) as in_f:
        bonzai_data = json.load(in_f)
    with open(BERT_DATA) as in_f:
        bert_data = json.load(in_f)

    conflicts = 0
    votes_nocrf = 0
    votes_nobonzai = 0
    votes_nobert = 0
    conflicts_heuristic = 0

    full_list = []
    for reg_i, crf_i, bonzai_i, bert_i in zip(regex_data, crf_data, bonzai_data, bert_data):
        res = dict()
        res['identifier'] = bonzai_i['identifier']

        # print(reg_i['text'].replace('\n', ' \\n '))
        # print(crf_i['text'].replace('\n', ' \\n '))
        # print(bonzai_i['text'].replace('\n', ' \\n '))
        res['text'] = bonzai_i['text']

        res['annotations'] = []

        for reg_j, crf_j, bonzai_j, bert_j in zip(reg_i["annotations"],
                                                  crf_i["annotations"],
                                                  bonzai_i["annotations"],
                                                  bert_i["annotations"]):
            token = dict()
            token['form'] = bonzai_j['form']
            token['begin'] = bonzai_j['begin']
            token['end'] = bonzai_j['end']

            if reg_j["label"] in CLASSES_REGEX:
                token['label'] = reg_j["label"]
            else:
                if len(res['text'].split('\n')) == 1 and bert_j['label_bert'] != 'O': # Si le texte ne contient pas de retour à la ligne, laisser BERT prendre la décision (excluant 'O')
                    token['label'] = bert_j['label_bert']
                if crf_j['label'] != bonzai_j['label_bonzaiboost'] or crf_j['form'] != bert_j['label_bert'] or bonzai_j['label_bonzaiboost'] != bert_j.get('label_bert', 'O'):
                    # Compte le nombre de fois où au moins un des réseaux n'est pas d'accord avec les autres
                    conflicts += 1


                if crf_j['label'] == bonzai_j['label_bonzaiboost']:  # Si CRF et BonzaiBoost votent pour la même classe, prendre celle-là
                    # Contient également tous les cas où les trois sont d'accord
                    token['label'] = crf_j['label']
                    if crf_j['label'] != bert_j['label_bert'] and bert_j['label_bert'] != 'O':
                        # Compte le nombre de fois où CRF et BonzaiBoost étaient d'accord contre l'avis de BERT, en dehors de 'O'
                        votes_nobert += 1
                elif crf_j['label'] == bert_j['label_bert']: # Si CRF et BERT votent pour la même classe, prendre celle-là
                    token['label'] = crf_j['label']
                    if crf_j['label'] != bonzai_j['label_bonzaiboost']:
                        # Compte le nombre de fois où CRF et BERT étaient d'accord contre l'avis de Bonzaiboost
                        votes_nobonzai += 1
                elif bonzai_j['label_bonzaiboost'] == bert_j['label_bert']: # Si BonzaiBoost et BERT votent pour la même classe, prendre celle-là
                    token['label'] = bonzai_j['label_bonzaiboost']
                    if crf_j['label'] != bonzai_j['label_bonzaiboost'] and crf_j['label'] != 'None':
                        # Compte le nombre de fois où BERT et BonzaiBoost étaient d'accord contre l'avis de CRF, en dehors de 'None'
                        votes_nocrf += 1
                else:
                    conflicts_heuristic += 1
                    print(res['text'].replace('\n', '\\n'))
                    if crf_j['form'] != bonzai_j['form'] or crf_j['form'] != bert_j['form']:
                        print(crf_j['form'], bonzai_j['form'], bert_j['form'])
                        raise Exception
                    print(crf_j['form'])
                    print(crf_j['label'], bonzai_j['label_bonzaiboost'], bert_j['label_bert'])
                    if bert_j['label_bert'] != 'O':
                        token['label'] = bert_j['label_bert']
                    elif crf_j['label'] != 'None':
                        token['label'] = crf_j['label']
                    else:
                        token['label'] = bonzai_j['label_bonzaiboost']
                    print(token['label'])

            # elif crf_j['label'] in CLASSES_CRF:
            #     token['label'] = crf_j['label']
            # elif bonzai_j['label_bonzaiboost'] in CLASSES_BONZAI:
            #     token['label'] = bonzai_j['label_bonzaiboost']
            # else:
            #     token['label'] = crf_j['label']

            res['annotations'] += [token]

        full_list += [res]

    print(conflicts, "conflicts resolved")
    print(votes_nocrf, 'votes against crf')
    print(votes_nobonzai, 'votes against bonzai')
    print(votes_nobert, 'votes against bert')
    print(conflicts_heuristic, "by heuristic")

    with open('preds_final.json', 'w') as out_f:
        out_f.write(json.dumps(full_list, ensure_ascii=False))


src/data/textmine.py
import json
from os import path
from typing import List
import pandas as pd

from torch.utils.data import Dataset
from transformers import PreTrainedTokenizer, CamembertTokenizerFast, AutoTokenizer

FILES = {
	'jdf': 'JDF.json',
	'jdr': 'JDR.json',
	'labels': 'label_idx.json'
}

CACHE_FILE = {
	'jdf': 'jdf.parquet',
	'jdr': 'jdr.parquet',
	'full': 'full.parquet',
	'train': 'train.parquet',
	'val': 'val.parquet',
	'test': 'test.parquet'
}

VAL_SPLIT = .15
TEST_SPLIT = .15

class TextMineDataset(Dataset):
	"""
	Textmine2022 dataset on recognizing named entity in signature
	"""
	
	def __init__(self,
	             split,
	             tokenizer: PreTrainedTokenizer=None,
	             labels:List[str] =None,
	             data_path:str=path.join('..', 'dataset'),
	             cache:str=path.join('..', '.cache')):
		"""
		Textmine2023 dataset
		:param split: precise the split to initialize dataset ('train', 'val', 'jdf', 'jdr')
		:type split: str
		:param tokenizer: hugging face tokenzier
		:type tokenizer: transformers.PreTrainedTokenizer
		:param labels: If given, will retains only these labels and ignore others
		:type labels: List[str]
		:param data_path: Where to find raw data
		:type data_path: str
		:param cache: path to export cache files
		:type cache: str
		"""
		self.split = split
		
		if isinstance(tokenizer, str):
			tokenizer = AutoTokenizer.from_pretrained(tokenizer, cache_dir=path.join(cache, 'transformers'))
		if tokenizer is None:
			tokenizer = CamembertTokenizerFast.from_pretrained("camembert-base", cache_dir=cache)
		self.tokenizer = tokenizer
		
		cache_path = path.join(cache, CACHE_FILE[split])
		label_path = path.join(cache, FILES['labels'])
		
		## Pre process data and load from cache files
		if path.exists(cache_path):
			self.data = pd.read_parquet(cache_path)
			self.data['input_ids'] = self.data['input_ids'].apply(lambda x: x.tolist())
			self.data['attention_mask'] = self.data['attention_mask'].apply(lambda x: x.tolist())
			self.data['ner_tags'] = self.data['ner_tags'].apply(lambda x: x.tolist())
			print(f'Load cache data from {cache_path}')
			
		else:
			
			with open(path.join(data_path, FILES[split]), 'r') as f:
				json_data = json.load(f)
			
			# clean data
			self.__remove_overlapping(json_data)
			self.__check_overlapping(json_data)
			
			# Tokenize text
			texts = [d['text'] for d in json_data]
			annotations = [d['annotations'] for d in json_data]
			
			token_encodings = tokenizer(texts, return_offsets_mapping=True)
			token_encodings['tokens'] = [tokenizer.convert_ids_to_tokens(input_ids) for input_ids in token_encodings['input_ids']]
			
			# Mapping labels
			token_span = token_encodings.offset_mapping
			token_span_dict = [[{'begin': span[0], 'end': span[1]} for span in token_sent] for token_sent in token_span]
			token_encodings['ner_tags'] = self.__mapping_nertag(token_span_dict, annotations)
			
			self.data = pd.DataFrame(dict(token_encodings))
			self.data.to_parquet(cache_path)
			print(f'Cache data at {cache_path}')
		
		## Load labels
		if path.exists(label_path):
			with open(label_path) as f:
				labels_idx = json.load(f)
				self.id2label = labels_idx['id2label']
				self.label2id = labels_idx['label2id']
				print(f'Load label idx from {label_path}')
		else:
			print(f'Generate new label2idx')
			unique_labels = list(set([i for l in self.data['ner_tags'] for i in l]))
			self.id2label = sorted(unique_labels, key=lambda x: 0 if len(x) == 1 else (ord(x[2]))*1e3 + ord(x[0]), reverse=False)
			self.label2id = {label: idx for idx, label in enumerate(self.id2label)}
		
		if labels is not None:
			# Convert to 'O' for all ner tag not in label list
			self.id2label = labels
			self.data['ner_tags'] = self.data['ner_tags'].apply(lambda ner_vector: ['O' if ner not in labels else ner for ner in ner_vector])
		
		self.data['labels'] = self.data['ner_tags'].apply(lambda x: [self.label2id[x_] for x_ in x])
	
	def __getitem__(self, idx):
		if idx >= len(self): raise IndexError  # meet the end of dataset
		sample = self.data.loc[idx].to_dict()
		sample = {k: sample[k] for k in ['input_ids', 'attention_mask', 'labels']}
		return sample
	
	def __len__(self):
		return len(self.data)
	
	@staticmethod
	def __remove_overlapping(data):
		for data_row in data:
			annotations = data_row['annotations']
			for pre, post in zip(annotations[:-1], annotations[1:]):
				if pre['end'] >= post['end']:
					annotations.remove(post)
	
	@staticmethod
	def __check_overlapping(data):
		for data_row in data:
			annotations = data_row['annotations']
			for pre, post in zip(annotations[:-1], annotations[1:]):
				if pre['end'] > post['begin']:
					print('TEXT:', data_row['text'])
					print(pre)
					print(post)
					print('=' * 15)
		
	
	@staticmethod
	def __mapping_nertag(token_span_batch, annotations_batch):
		"""
		    Remap IOB tag to each token generated by tokenizer. Should provide the span (begin/end)
		    """
		
		labels = list()
		
		for token_span_sent, annotations in zip(token_span_batch, annotations_batch):
			
			annotations = annotations.copy()
			entity = annotations.pop(0)
			
			last_label = 'O'
			token_label = list()
			
			for token in token_span_sent:
				
				while entity['end'] < token['begin']: entity = annotations.pop(0)
				
				if token['begin'] == token['end']:
					label = 'O'
				elif entity['begin'] <= token['begin'] and token['end'] <= entity['end']:
					prefix = 'B-' if last_label == 'O' or last_label[2:] != entity['label'] else 'I-'
					label = prefix + entity['label']
				else:
					label = 'O'
				
				token_label.append(label)
				last_label = label
			
			labels.append(token_label)
		
		return labels


models/graph_encoder/main.py
import click
import pandas as pd
import numpy as np
from sklearn.metrics import f1_score

from os import getenv 

from rich.progress import track 
from libraries.log import logger 
from libraries.strategies import *
from data_holder import GraphDataset 
from modelization import TextMineGCN

from torch_geometric.loader import DataLoader
import pytorch_lightning as pl
from pytorch_lightning.loggers import TensorBoardLogger
from pytorch_lightning.callbacks import ModelCheckpoint

@click.group(chain=False, invoke_without_command=True)
@click.pass_context
def router(ctx):
    ctx.ensure_object(dict)
    logger.debug('nltk initialization ...!')

    path2models = getenv('MODELS')
    path2dataset = getenv('DATASET')
    path2features = getenv('FEATURES')
    path2nltk_data = getenv('NLTK_DATA')
    path2checkpoints = getenv('CHECKPOINTS')
    
    logger.debug('env variables loading')

    ctx.obj['path2models'] = path2models
    ctx.obj['path2dataset'] = path2dataset 
    ctx.obj['path2features'] = path2features  
    ctx.obj['path2nltk_data'] = path2nltk_data 
    ctx.obj['path2checkpoints'] = path2checkpoints

    logger.debug('nltk initialization')
    nltk.download('punkt', download_dir=path2nltk_data) 
    nltk.download('stopwords', download_dir=path2nltk_data)
    
    ctx.ensure_object(dict)
    subcommand = ctx.invoked_subcommand
    if subcommand is None:
        logger.debug('use --help option to see availables options')
    else:
        logger.debug(f'{subcommand} was called')

@router.command()
@click.option('--vectorizer', help='vectorizer name(feature extraction)', type=str)
@click.option('--corpus_filename', help='corpus file name', type=str)
@click.option('--graph_filename', help='features serialization file name', type=str)
@click.pass_context
def preprocessing(ctx, vectorizer, corpus_filename, graph_filename):
    try:
        device = th.device('cuda:0' if th.cuda.is_available() else 'cpu')

        path2models = ctx.obj['path2models']
        path2dataset = ctx.obj['path2dataset']
        path2features = ctx.obj['path2features']

        path2corpus = path.join(path2dataset, corpus_filename)
        if '/' in vectorizer:
            path2vectorizer = path.join(path2models, ''.join(vectorizer.split('/')))
        else:
            path2vectorizer = path.join(path2models, vectorizer)
        path2_graphs = path.join(path2features, graph_filename)

        tokenizer, vectorizer = load_vectorizer(path2vectorizer, vectorizer)
        embedding_model = vectorizer.embeddings
        corpus = load_dataset(path2corpus)
        with open(path.join(path2dataset, 'label2id.json'), 'r') as f:
            label2id = json.load(f)

        global_graph_accumulator = []
        documents = corpus['text'].tolist()
        labels = corpus['annotations'].tolist()
        ids = corpus['identifier'].tolist()
        # For each document in corpus
        for doc_id, doc in enumerate(track(documents, f'processing graphs...')):
            toks, token_pos, token_line = align_tokens(doc, labels[doc_id])
            tokens = tokenizer(toks, return_tensors='pt', return_offsets_mapping=True, is_split_into_words=True)
            tokens, offsets_mapping = {k:v for k,v in tokens.items() if k != 'offset_mapping'}, tokens['offset_mapping']

            embeddings_matrix = embedding_model(tokens['input_ids']).cpu()

            graph = build_graph(doc, embeddings_matrix[0,:,:], 2, labels[doc_id], ids[doc_id], offsets_mapping, token_line, token_pos, label2id)
            global_graph_accumulator.append(graph)
        serialize(path2_graphs, global_graph_accumulator, pickle)
        logger.success(f'{len(global_graph_accumulator):03d} dgl graphs were saved...!')
    except Exception as e:
        logger.error(e)
    
@router.command()
@click.option('--graph_filename', help='', type=str)
@click.option('--valgraph_filename', help='', type=str)
@click.option('--nb_epochs', type=int, default=32)
@click.option('--batch_size', type=int, default=16)
@click.option('--checkpoint', type=str)
@click.option('--period', default=8, type=int)
@click.pass_context
def train(ctx, graph_filename, valgraph_filename, nb_epochs, batch_size, checkpoint, period):
    
    device = th.device('cuda:0' if th.cuda.is_available() else 'cpu')
    logger.debug(f'target device : {device}')

    path2features = ctx.obj['path2features']
    path2checkpoints = ctx.obj['path2checkpoints']

    path2graph_filename = path.join(path2features, graph_filename)
    path2valgraph_filename = path.join(path2features, valgraph_filename)
    
    train_dataset = GraphDataset(path2graph_filename, begin_percentage=0.0, end_percentage=0.75)
    val_dataset = GraphDataset(path2valgraph_filename, begin_percentage=0.75, end_percentage=100.0)
    train_dataloader = DataLoader(train_dataset, batch_size, shuffle = True)
    val_dataloader = DataLoader(val_dataset, batch_size, shuffle=False, num_workers=8)
    
    logger.debug('dataset and dataloader were initialized')

    path2initial_weights = path.join(path2checkpoints, checkpoint)
    if path.isfile(path2initial_weights):
        logger.debug(f'initial weights were detected, the model will use {path2initial_weights}')
        model = TextMineGCN.load_from_checkpoint(path2initial_weights, dataset=train_dataset)
    else:
        logger.debug('no initial weights were detected, the model will be created from scratch')
        model = TextMineGCN(train_dataset)
        model.to(device)

    print(model)
    logger.debug('model was initialized')

    logger.debug('training will start in 1 s')
    
    training_logger = TensorBoardLogger(path2initial_weights, name="tb_logs", version=0)

    print(model)
    logger.debug('model was initialized')
    callback = ModelCheckpoint(dirpath=training_logger.log_dir, save_top_k=3, monitor='val_loss', mode='min')
    trainer = pl.Trainer(max_epochs=nb_epochs, logger=training_logger, enable_checkpointing=True, default_root_dir=training_logger.log_dir, track_grad_norm=-1)
    logger.debug('training will start in 1 s')

    trainer.fit(model, train_dataloader, val_dataloader)
    
    path2model_weights = path.join(path2checkpoints, 'checkpoint_###.th')
    th.save(model.cpu(), path2model_weights)
    logger.success(f'training end | weights are saved at {path2model_weights}')

@router.command()
@click.option('--graph_filename', help='', type=str)
@click.option('--batch_size', type=int, default=16)
@click.option('--checkpoint', type=str)
@click.pass_context
def predict(ctx, graph_filename, batch_size, checkpoint):
    
    device = th.device('cuda:0' if th.cuda.is_available() else 'cpu')
    logger.debug(f'target device : {device}')

    path2features = ctx.obj['path2features']
    path2checkpoints = ctx.obj['path2checkpoints']

    path2graph_filename = path.join(path2features, graph_filename)
    
    dataset = GraphDataset(path2graph_filename, begin_percentage=0.0, end_percentage=0.75)
    dataloader = DataLoader(dataset, batch_size, shuffle = True, num_workers=8)
    
    logger.debug('dataset and dataloader were initialized')

    path2initial_weights = path.join(path2checkpoints, checkpoint)
    model = TextMineGCN.load_from_checkpoint(path2initial_weights, dataset=dataset)

    print(model)
    logger.debug('model was initialized')

    for batch in dataloader:
        

if __name__ == '__main__':
    router(obj={})


models/graph_encoder/modelization.py
import numpy as np 
import torch as th 
import torch.nn.functional as F
import torch.nn as nn 
from torch_geometric.nn import GCNConv, global_mean_pool
import pytorch_lightning as pl
from torchmetrics import F1Score

import operator as op
import itertools as it, functools as ft 

class PositionalEncoding(nn.Module):
    def __init__(self, max_nb_nodes, in_dim, drop_val=0.1):
        super(PositionalEncoding, self).__init__()
        pos = np.arange(0, max_nb_nodes)[:, None]
        idx = np.fromfunction(lambda _,j: j - j % 2, shape=(1, in_dim))
        mask = np.fromfunction(lambda _,j: j % 2 == 0, shape=(1, in_dim))
        pnt = pos / (10000 ** (idx / in_dim))
        val = np.sin(pnt) * mask + np.cos(pnt) * (1 - mask)
        self.drop_layer = nn.Dropout(drop_val)
        self.register_buffer('psne_layer', th.tensor(val).float())

    def forward(self, node_positions):
        pos = self.psne_layer[node_positions, :]
        return self.drop_layer(pos)

class NRLModel(nn.Module):
    def __init__(self):
        super(NRLModel, self).__init__()
        self.activation_map = {
            0: nn.Identity(), 
            1: nn.ReLU(), 
            2: nn.LeakyReLU(0.2),
            3: nn.GELU(), 
            4: nn.Tanh(), 
            5: nn.Sigmoid(), 
            6: nn.Softmax(dim=-1)
        }

class MLPModel(NRLModel):
    def __init__(self, layers_config, activations, apply_norm=False):
        super(MLPModel, self).__init__()

        assert isinstance(layers_config, list)
        assert isinstance(activations, list)
        assert len(layers_config)  - 1 == len(activations) 

        self.shapes = list(zip(layers_config[:-1], layers_config[1:]))
        self.activations = activations 
        self.linear_layers = nn.ModuleList([]) 
        
        for index, (in_dim, out_dim) in enumerate(self.shapes):
            linear = nn.Linear(in_features=in_dim, out_features=out_dim)
            nonlinear = self.activation_map[self.activations[index]] 
            if apply_norm and index < len(self.shapes) - 1:
                normalizer = nn.BatchNorm1d(out_dim)
                layer = nn.Sequential(linear, normalizer, nonlinear)
            else:
                layer = nn.Sequential(linear, nonlinear)
            self.linear_layers.append(layer)
    
    def forward(self, X_0):
        X_N = ft.reduce(lambda X_I, LIN: LIN(X_I), self.linear_layers, X_0)
        return X_N 
    

# class GCNModel(NRLModel):
#     def __init__(self, layers_config, activations, drop_val=0.1, acc='mean'):
#         super(GCNModel, self).__init__()

#         assert isinstance(layers_config, list)
#         assert isinstance(activations, list)
#         assert len(layers_config) - 1 == len(activations) 

#         self.acc = acc 
#         self.shapes = list(zip(layers_config[:-1], layers_config[1:]))
#         self.activations = activations 
#         self.convolutions = nn.ModuleList([])
        
#         for index, (in_dim, out_dim) in enumerate(self.shapes):
#             fun = self.activation_map.get(self.activations[index], nn.Identity()) 
#             layer = dglnn.SAGEConv(
#                 in_feats=in_dim, 
#                 out_feats=out_dim, 
#                 aggregator_type=self.acc,
#                 activation=fun,
#                 feat_drop=drop_val
#             )
#             self.convolutions.append(layer)
    
#     def forward(self, G, X_0, E_0):
#         X_N = ft.reduce(
#             lambda X_I, GCN: GCN(G, X_I, E_0), 
#             self.convolutions, 
#             X_0 
#         )
#         with G.local_scope():
#             G.ndata['X_N'] = X_N 
#             graph_embedding = dgl.readout_nodes(G, 'X_N')
#             return graph_embedding

# class GNNModel(nn.Module):
#     def __init__(self, config):
#         super(GNNModel, self).__init__()
#         self.fst_gcn_vectorizer = GCNModel(**config)
#         self.snd_gcn_vectorizer = GCNModel(**config)
    
#     def forward(self, GF, GS, XF_0, XS_0, EF_0=None, ES_0=None):
#         XF_N = self.fst_gcn_vectorizer(GF, XF_0, E_0=EF_0)
#         XS_N = self.snd_gcn_vectorizer(GS, XS_0, E_0=ES_0)
#         return XF_N, XS_N

#     def graph_embedding(self, G, X_0, E_0):
#         return self.fst_gcn_vectorizer(G, X_0, E_0).cpu().numpy()

class SimpleGCN(th.nn.Module):
    def __init__(self, dataset):
        super().__init__()
        self.conv1 = GCNConv(dataset.num_node_features, 756)
        self.conv2 = GCNConv(756, dataset.num_classes)

    def forward(self, data):
        x, edge_index = data.embedding, data.edge_index

        x = self.conv1(x, edge_index)
        x = F.relu(x)
        x = F.dropout(x, training=self.training)
        x = self.conv2(x, edge_index)

        return global_mean_pool(x, data.batch)

class TextMineGCN(pl.LightningModule):
    def __init__(self, dataset):
        super().__init__()
        self.conv1 = GCNConv(dataset.num_node_features, 756)
        self.conv2 = GCNConv(756, dataset.num_classes)
        self.criterion = nn.CrossEntropyLoss()
        self.f1 = F1Score(task="multiclass", num_classes=dataset.num_classes)

    def training_step(self, batch, batch_idx):
        batch_size = len(batch)
        label = th.tensor(list(it.chain(*batch.label)), device = self.device)
        label_pos = list(it.chain(*batch.label_pos))
        x, edge_index = batch.embedding, batch.edge_index
        x = self.conv1(x, edge_index)
        x = F.relu(x)
        x = F.dropout(x, training=self.training)
        x = self.conv2(x, edge_index)
        pred = th.argmax(x, dim=-1)
        loss = self.criterion(x[label_pos], label)
        # Logging to TensorBoard by default
        self.log("train_loss", loss, batch_size=batch_size)
        return loss
    
    def validation_step(self, val_batch, batch_idx):
        batch_size = len(val_batch)
        label = th.tensor(list(it.chain(*val_batch.label)), device=self.device)
        label_pos = list(it.chain(*val_batch.label_pos))
        x, edge_index = val_batch.embedding, val_batch.edge_index
        x = self.conv1(x, edge_index)
        x = F.relu(x)
        x = F.dropout(x, training=self.training)
        x = self.conv2(x, edge_index)
        pred = th.argmax(x, dim=-1)
        loss = self.criterion(x[label_pos], label)
        # Logging to TensorBoard by default
        self.log("val_loss", loss, prog_bar=True, batch_size=batch_size)
        f1 = self.f1(pred[label_pos], label)
        self.log('val_f1', f1, prog_bar=True, batch_size=batch_size)
        return loss
    
    def predict(self)

    def configure_optimizers(self):
        optimizer = th.optim.Adam(self.parameters(), lr=1e-3)
        return optimizer

processing_bonzaiboost.py
import subprocess

import numpy as np
import pandas as pd
import json
import re
import sys
from subprocess import run
from tqdm import tqdm

INPUT_DATA = ['dataset/JDR.json',
              'dataset/JDF.json']

FOLDER = './bonzai_data/'
STEM = 'textmine'

N_ESTIMATORS = '200'
OVERSAMPLING_CLASSES = {'Email': 4,
                        'Function': 4,
                        'Phone_Number': 2,
                        'Project': 15,
                        'Reference_CEDEX': 15,
                        'Reference_CS': 30,
                        'Reference_Code_Postal': 3,
                        'Reference_User': 30,
                        'Social_Network': 100,
                        'Url': 8}

CLASSES = ['Human', 'Organization', 'Function', 'Project', 'Location', 'Reference_CEDEX', 'Reference_CS',
           'Reference_Code_Postal', 'Phone_Number', 'Email', 'Url', 'Social_Network', 'Reference_User']
BERT_CLASSES = []
# BERT_CLASSES = ['Project', 'Organization', 'Function', 'Location']

FEATURES = {'token': 'text',
            'token_split': 'text',
            'token_lower': 'text',
            'before_on_line': 'text',
            'before_on_line_lower': 'text',
            'before': 'text',
            'after_on_line': 'text',
            'after_on_line_lower': 'text',
            'after': 'text',
            # 'line': 'continuous'
            }

REGEX_M = {
    'mail': r'[\w\-\'\.]+@[\w\-]+[.]\w+',
    'tel': r'(?:[+]\D?)?\d+(?:\D{0,2}\d){6,12}',
    'url': r'[\w\-\.]+[.][a-zA-Z]+',
    'http': r'http.*',
    'handle': r'^@.*',
    'postcode': r'\d{5}',
    'caps': r'^[A-Z]',
    'allcaps': r'^[A-Z]+$',
    'nonword': r'^[\d\W]+$',
    'social': r'Facebook|Linkedin|Instagram|Twitter|Google|Youtube'
}
REGEX_T = {
    'suffix_': r'.?.?.?$'
}
TEXT_PARAMS = {'expert_length': 1,
               'expert_type': 'ngram',
               'cutoff': 0
               }

EVAL_DATA = 'dataset/JDA.json'


def cleanup(s):
    return s.replace('\n', ' \\n ')


def regex_string(name, regex, mode="M", level='0'):
    return 'regex=' + name + '_%' + mode + '[' + level + ',"' + str(regex) + '"]'


def create_names(filename):
    with open(filename, 'w') as f:
        labels_activated = [i for i in CLASSES if i not in BERT_CLASSES]
        if len(BERT_CLASSES) != 0:
            labels_activated += ['BERT']
        f.write(','.join(labels_activated) + '.\n')
        for i in FEATURES:
            if FEATURES[i] == 'text':
                param_str = ' '.join([i + '=' + str(TEXT_PARAMS[i]) for i in TEXT_PARAMS])
                regex_str = ' '.join([regex_string(i, REGEX_M[i], 'M') for i in REGEX_M] +
                                     [regex_string(i, REGEX_T[i], 'T') for i in REGEX_T]
                                     )
                f.write(i + ': text: ' + param_str + ' ' + regex_str + '.\n')
            elif FEATURES[i] == 'continuous':
                f.write(i + ': continuous.\n')


def process_features(token, text):
    # text = text.replace('\n', ' ')

    begin = token['begin']
    end = token['end']

    before = text[:begin]
    after = text[end:]

    before_on_line = before.split('\n')[-1] if len(before) != 0 else ""
    after_on_line = after.split('\n')[0] if len(after) != 0 else ""

    features = {'token': token['form'].replace('\n', ''),
                'token_split': ' '.join(re.split('(\W)', token['form'].replace('\n', ''))),
                'before': cleanup(before),
                'after': cleanup(after),
                'before_on_line': before_on_line,
                'after_on_line': after_on_line,
                'token_lower': token['form'].replace('\n', '').lower(),
                'before_on_line_lower': before_on_line.lower(),
                'after_on_line_lower': after_on_line.lower(),
                'line': str(len(re.findall(r'\n', before)))
                }

    return features


def process_data_train(data, out_file):
    for signature in data:
        text = signature['text']  # .replace('\n', ' \\n ')
        for token in signature['annotations']:
            features = process_features(token, text)

            label = token['label']

            oversampling = OVERSAMPLING_CLASSES.get(label, None)
            oversampling_marker = str(oversampling) + '::' if oversampling is not None else ''

            label_show = label if label not in BERT_CLASSES else 'BERT'

            out_file.write(oversampling_marker + '?'.join([features[i] for i in FEATURES] + [label_show]) + '.\n')


def process_data_predict(data, out_file):
    with open('bonzai_data/predict.log', 'w') as f:
        f.write('')
    for signature in tqdm(data):
        text = signature['text']  # .replace('\n', ' \\n ')
        for token in signature['annotations']:
            features = process_features(token, text)

            line = '?'.join([features[i] for i in FEATURES] + ['']) + '.'

            result = run([
                "bonzaiboost -S " + STEM + " -boost adamh -n " + N_ESTIMATORS + " --sep '?' -C -c single -o backoff -v 0 2> " + 'predict_err.log'],
                shell=True, cwd=FOLDER, stdout=subprocess.PIPE, input=line.encode(), stderr=subprocess.STDOUT)
            pipe_out = result.stdout.decode("utf-8")
            with open('bonzai_data/predict.log', 'a') as f:
                f.write(line + '\n' + pipe_out + '\n')
            prediction = pipe_out.split('->')[1].strip(' []\n')
            token['label_bonzaiboost'] = prediction

    outs = json.dumps(data, ensure_ascii=False)
    out_file.write(outs)


TRAIN = True
EVAL = True
PRED = False

if __name__ == '__main__':
    create_names(FOLDER + STEM + '.names')

    if TRAIN:
        print("Training")
        for i, filename in enumerate(INPUT_DATA):
            with open(filename) as in_f:
                print('Loading', filename)
                data = json.load(in_f)
            open_mode = 'w' if i == 0 else 'a'
            with open(FOLDER + STEM + '.data', open_mode) as out_f:
                process_data_train(data, out_f)

        if EVAL:
            # Run cross-validation for an evaluation of the model
            run([
                "bonzaiboost -S " + FOLDER + STEM + " -boost adamh -n " + N_ESTIMATORS + " --sep '?' -cross k10 -jobs 4 -c single -o backoff > " + FOLDER + "crossval.log"],
                shell=True)

        # Train one simple model to be able to extract its features
        run(["bonzaiboost -S " + STEM + " -boost adamh -n " + N_ESTIMATORS + " --sep '?'"], cwd=FOLDER, shell=True)
        run(["bonzaiboost -S " + STEM + " -boost adamh -n " + N_ESTIMATORS + " --sep '?' --info -o 0.005"], cwd=FOLDER,
            shell=True)

    if PRED:
        print("Predictions")
        with open(EVAL_DATA) as in_f:
            print('Loading', EVAL_DATA)
            data = json.load(in_f)
        with open(FOLDER + 'preds.json', 'w', encoding='utf-8') as out_f:
            process_data_predict(data, out_f)
